Improved page mapping benchmarks

alwin-joshy opened the pull request:
The existing page mapping benchmarks only appear to be implemented for x86 platforms.  In this PR, I have extended the benchmarks to work on AARCH64 and fixed a bug with the protect benchmark, which would result in the same page being protected repeatedly instead of protection being applied to all of the pages in the range. I have also modified the CMake to be more explicit that this benchmark is currently only supported on x86 and AARCH64.

Because of the changes to vspace mapping in https://github.com/seL4/seL4/pull/801, the logic for preparing the page tables is quite different on AARCH64 compared to x86_64, which retains the design of unique kernel objects for each level of the page table. Because it is not obvious if a page table mapping succeeding means that the lowest level page table required for mapping a small page exists, I repeatedly map page tables for an address until the mapping operation fails with `seL4_DeleteFirst`, indicating that the page table has already been completely filled in for this address. Not sure if this is the ideal way of doing this and I can change it to do something else if not.

alwin-joshy committed:
Improve page mapping benchmarks

Extended page mapping benchmarks to work on AARCH64,
changed CMake to make it clearer which architectures
are supported, and fixed a bug in the implementation.

Signed-off-by: Alwin Joshy <joshyalwin@gmail.com>

axel-h commented on apps/page_mapping/src/main.c:
```diff
@@ -18,9 +18,11 @@
 
 #if defined(CONFIG_ARCH_X86_64)
 #define DEFAULT_DEPTH 64
+#elif defined(CONFIG_ARCH_AARCH64)
```
Can't we simply use
```
#define DEFAULT_DEPTH (sizeof(seL4_Word)*8)
```

axel-h commented on apps/page_mapping/src/main.c:
```diff
@@ -186,17 +206,28 @@ bench_proc(int argc UNUSED, char *argv[])
+                ZF_LOGF_IFERR(err, "ummap page table\n");
+            }
+        }
+#else
```
better make this more specific
```
#elif defined(CONFIG_ARCH_AARCH64)
```

axel-h replied:
So it could be this:
```suggestion
#ifdef CONFIG_ARCH_X86_64
        err = seL4_X86_PDPT_Unmap(i);
        if (!err) { continue; }
        err = seL4_X86_PageDirectory_Unmap(i);
        if (!err) { continue; }
#endif /* CONFIG_ARCH_X86_64 */
        err = seL4_ARCH_PageTable_Unmap(i);
        ZF_LOGF_IFERR(err, "ummap page table\n");
```

alwin-joshy replied:
That would work, though I'm not sure it helps much in terms of clarity.

axel-h commented on apps/page_mapping/src/main.c:
```diff
@@ -77,8 +78,23 @@ static void inline prepare_page_table(seL4_Word addr, int npage, seL4_CPtr untyp
+            assert(err == 0);
+            (*free_slot)++;
+
+            err = seL4_ARCH_PageTable_Map(pt, SEL4UTILS_PD_SLOT, addr,
```
Should'n there be an `assert(err == 0);` afterwards then also, just to play safe?

axel-h commented on apps/page_mapping/CMakeLists.txt:
```diff
@@ -16,7 +16,7 @@ config_option(
     ON
     DEPENDS
-    "DefaultBenchDeps"
+    "DefaultBenchDeps; KernelSel4ArchAarch64 OR KernelArchX86"
```
What about RISC-V now actually?

alwin-joshy replied:
Yeah, I was thinking that too. Since RISC-V uses pretty much the same kind of page table structure as AARCH64, it would probably work with these changes. Do you think that's worth adding to this PR?

axel-h replied:
I you have the time to check RISC-V also, that would be good. Seems we are just excluding arm32 then.

alwin-joshy replied:
Finally had some time to revisit this, I have confirmed the benchmarks work on RISC-V (though I haven't run the full benchmark with up to 2048 pages and 16 runs per configuration to completion as it takes forever).

alwin-joshy committed:
Page mapping benchmarks: Add RISC-V support

This commit enables page-mapping benchmarks for RISC-V. No changes were
needed due to the changes to AARCH64 page table structures, which made
each level of the page table a generic PageTable object (which was
already the case in RISC-V).

Signed-off-by: Alwin Joshy <joshyalwin@gmail.com>

lsf37 committed:
Merge branch 'master' into improve_page_map
